dot/parachain/inherent.go

```
@@ -0,0 +1,157 @@
+package parachain
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/gossamer/internal/log"
+
+	provisioner "github.com/ChainSafe/gossamer/dot/parachain/provisioner/messages"
+	parachaintypes "github.com/ChainSafe/gossamer/dot/parachain/types"
+	"github.com/ChainSafe/gossamer/dot/types"
+	"github.com/ChainSafe/gossamer/lib/common"
+)
+
+var logger = log.NewFromGlobal(log.AddContext("pkg", "parachain"))
+
+var errResponseChannelClosed = fmt.Errorf("response channel is closed")
+
+// InherentData represents parachains inherent-data passed into the runtime by a block author
+type InherentData struct {
+	// Bitfields represents signed bitfields by validators about availability.
+	Bitfields []parachaintypes.UncheckedSignedAvailabilityBitfield `scale:"1"`
+	// BackedCandidates represents backed candidates for inclusion in the block.
+	BackedCandidates []parachaintypes.BackedCandidate `scale:"2"`
+	// Disputes represents sets of dispute votes for inclusion.
+	Disputes []parachaintypes.DisputeStatementSet `scale:"3"`
+	// ParentHeader represents the parent block header. Used for checking state proofs.
+	ParentHeader types.Header `scale:"4"`
+}
+
+type blockState interface {
+	GetHeader(common.Hash) (*types.Header, error)
+}
+
+func ProvideInherentData(
+	blockState blockState,
+	overseerCh chan<- any,
+	relayParent common.Hash,
+	inherentData *types.InherentData,
+) error {
+	parachainInherentData, err := createInherentData(blockState, overseerCh, relayParent)
+	if err != nil {
+		return fmt.Errorf("creating parachain inherent data: %w", err)
+	}
+
+	if inherentData == nil {
```

This [won't work](https://go.dev/play/p/65j1xqxZNmp) unless `inherentData` is a pointer to a pointer.

I think we should return an error instead.

---

interesting. Let's just return the inherent data and let the caller decide how it wants to use.

---

That's even better. "out parameters" are a bit of a code smell IMHO.